File: cpp/easy/longest_common_prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string prefix;
        string compare;
        if(strs.size() == 0)
            return prefix;
        if(strs.size() == 1)
            return strs[0];
        prefix = strs[0];
        int j = 0;
        for(int i = 1; i < strs.size(); ++i)
        {
            j = 0;
            while(j < prefix.size() && j < strs[i].size())
            {
                if(prefix[j] == strs[i][j])
                {
                    compare.push_back(prefix[j]);
                    ++j;
                }
                else
                    break;
            }
            prefix = compare;
            compare.clear();
        }
        return prefix;
    }
};


Conclusion:
easy algorithm issue.
Time complexity is: O(n*L), L is the lenght of common prefix
Space complexity is: O(1)

File: python/easy/longest_common_prefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        prefix = str()
        compare = str()
        if(len(strs) == 0):
            return prefix
        elif(len(strs) == 1):
            return strs[0]
        prefix = strs[0]
        j = 0
        for i in xrange(1, len(strs)):
            j = 0
            while(j < len(prefix) and j < len(strs[i])):
                if(prefix[j] == strs[i][j]):
                    compare += strs[i][j]
                else:
                    break
                j += 1
            prefix = compare
            compare = ""
        return prefix

# pythonic way
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        for idx, chars in enumerate(zip(*strs)):
            if len(set(chars)) > 1:
                return strs[0][:idx]
        return min(strs, key=len)
        
